test: Add bitcoin-chainstate test for assumeutxo functionality

stringintech opened the pull request:
This PR adds functional test coverage for the bitcoin-chainstate tool loading a datadir initialized with an assumeutxo snapshot.

The PR also includes:
- Fix for assertion crash in `ChainstateManager::AddChainstate()` when `prev_chainstate` has no initialized mempool (required for the test to pass)
- `-regtest` flag support for bitcoin-chainstate to enable the testing

This work started while experimenting with the bitcoin-chainstate tool and how the kernel API (#30595) behaved when loading a datadir containing assumeutxo data, during the time that PR was still under review. sedited suggested opening a PR to add this test coverage.

stringintech committed:
Add regtest support to bitcoin-chainstate tool

Adds -regtest flag to enable testing with regtest chain parameters.

stringintech committed:
test: Add bitcoin-chainstate test for assumeutxo functionality

Adds functional test coverage for bitcoin-chainstate tool loading a datadir initialized with an assumeutxo snapshot

frankomosh commented on src/validation.cpp:
```diff
@@ -6321,7 +6321,7 @@ Chainstate& ChainstateManager::ActivateExistingSnapshot(uint256 base_blockhash)
 
     // Mempool is empty at this point because we're still in IBD.
-    Assert(m_active_chainstate->m_mempool->size() == 0);
+    Assert(!m_active_chainstate->m_mempool || m_active_chainstate->m_mempool->size() == 0);
```
Is this now weaker that the original assertion, because I think it will now silently accepts a null mempool pointer?

TheCharlatan replied:
It is weaker, yes, but we similarly guard in other situations. We also still have the assertion in ActivateSnapshot to guard this condition at a more critical point in time.

frankomosh reviewed (commented):
Concept ACK
I think its a nice addition as it adds valuable functional coverage for bitcoin-chainstate.
Few nits and inquiries inline;

sedited reviewed (commented):
Concept ACK

Consolidating the tests into the regtest case seems fine to me, especially since this is just a utility to showcase and doogfood some of the kernel work.

TheCharlatan commented on src/bitcoin-chainstate.cpp:
```diff
@@ -49,10 +49,11 @@ int main(int argc, char* argv[])
+    if (argc < 2 || argc > 3) {
         std::cerr
-            << "Usage: " << argv[0] << " DATADIR" << std::endl
+            << "Usage: " << argv[0] << " DATADIR [-regtest]" << std::endl
```
Ah, I see. I think I would change the order then, since typically named optional args precede positional arguments / operands in command line applications: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_01 .

stringintech commented:
986eb77 to 3e7d272 - reversed the order of `-regtest` flag and input `DATADIR` as [suggested](https://github.com/bitcoin/bitcoin/pull/33728#discussion_r2475582408) by @TheCharlatan and added simple validation for the flag as [suggested](https://github.com/bitcoin/bitcoin/pull/33728#discussion_r2471886228) by @frankomosh.

sedited reviewed (approved):
ACK 3e7d272b2747f1937010b8fc1fefafcfa0ce3213

yuvicc commented:
Concept ACK

yuvicc commented on src/bitcoin-chainstate.cpp:
I think we can add here 2022 - present

frankomosh commented:
ACK 3e7d272

sedited reviewed (approved):
re-ACK 67740dfe6cfe3f3f91aaad41018a7da9004dcd8c

frankomosh commented:
re-ACK 67740df

sedited reviewed (approved):
Re-ACK 6cb63db8a97b8ea1dbee5edaff30e6da8b6ed108

sedited commented on src/bitcoin-chainstate.cpp:
```diff
@@ -131,16 +131,18 @@ class TestKernelNotifications : public KernelNotifications
     // SETUP: Argument parsing and handling
-    if (argc != 2) {
+    const bool has_regtest_flag{argc == 3 && std::string(argv[1]) == "-regtest"};
+    if (argc < 2 || argc > 3 || (argc == 3 && !has_regtest_flag)) {
```
> Makes sense, thanks! It seems I can cherry-pick your commit as is to replace https://github.com/bitcoin/bitcoin/commit/bfceb4c275f7df403d7fbc2e114934988cc8111c?

Sure, if you think it is a good idea too :)

stringintech committed:
Fix `ChainstateManager::AddChainstate()` assertion crash

Check mempool exists before accessing size when prev_chainstate doesn't have initialized mempool.

theStack reviewed (approved):
Concept and code-review ACK 7b5d256af4a0f954a919604ed4346db3a814fb6d

As a follow-up idea, it might make sense to deduplicate the deterministic generation of the snapshot chain between this test and feature_assumeutxo.py, by moving the logic into a test framework module.

sedited reviewed (approved):
Re-ACK 7b5d256af4a0f954a919604ed4346db3a814fb6d

achow101 commented:
ACK 7b5d256af4a0f954a919604ed4346db3a814fb6d
